### PR TITLE
Bluetooth: audio: ascs: Fix ASE state transition due to CIS link loss

### LINKED ISSUE #60669: LE Audio: Host: Wrong state after link loss of a CIS
The PTS BAP/USR/SCC/BV-168-C test case fails because at sudden link loss of CIS, the IUT (in server role) acts according to Table 3.2: 'ASE state machine transitions', which are valid for ASE Control Operations, but does not include the exception described in ASCS specification (ACSC_v1.0-1.pdf, 3.2 ASE state machine transitions):

'If the server detects link loss of a CIS for an ASE in the Streaming state or the Disabling state, the server shall immediately transition that ASE to the QoS Configured state. Link loss of a CIS for an ASE in any state other than Streaming or Disabling shall not cause a transition of the ASE state machine.'

So, even if ASE is Source, it should end up in QoS Configured, not in Disabling state.
https://github.com/zephyrproject-rtos/zephyr/blob/b69cd3cbc101cae3558a4730fa320e25a43b1543/subsys/bluetooth/audio/ascs.c#L898-L912

PTS logs:
[BAP_USR_SCC_BV_168_C_2023_07_21_09_15_30.zip](https://github.com/zephyrproject-rtos/zephyr/files/12117590/BAP_USR_SCC_BV_168_C_2023_07_21_09_15_30.zip)

IUT logs:
[BAP_USR_SCC_BV-168-C_iutctl.log](https://github.com/zephyrproject-rtos/zephyr/files/12117582/BAP_USR_SCC_BV-168-C_iutctl.log)

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This fixes ASE state transition to QoS Configured state that shall be
done autononously when CIS has been terminated due to link loss.

As per ASCS v1.0:
"If the server detects link loss of a CIS for an ASE in the Streaming
state or the Disabling state, the server shall immediately transition
that ASE to the QoS Configured state. Link loss of a CIS for an ASE in
any state other than Streaming or Disabling shall not cause a
transition of the ASE state machine."

The PR adds test cases testing the changes made:
test_cis_link_loss_in_streaming_state
test_cis_link_loss_in_disabling_state
test_cis_link_loss_in_enabling_state
test_cis_estab_failed_in_enabling_state

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/60669

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
Updates since the last revision:
* Improved the logic
* The non-link loss CIS disconnection triggers the autonomous Release operation now
* Added test case testing the above behavior

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/ascs.c [hunk 1]
```diff
@@ -106,6 +106,11 @@ NET_BUF_SIMPLE_DEFINE_STATIC(ase_buf, ASE_BUF_SIZE);
 static int control_point_notify(struct bt_conn *conn, const void *data, uint16_t len);
 static int ascs_ep_get_status(struct bt_bap_ep *ep, struct net_buf_simple *buf);
 
+static bool check_cis_link_loss(struct bt_ascs_ase *ase)
+{
+	return ase->ep.reason == BT_HCI_ERR_CONN_TIMEOUT;
```
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c:111]: Link loss happens when the remote part stops sending PDUs for a specific time (supervision timeout). The error codes you've mentioned mean:
`BT_HCI_ERR_REMOTE_USER_TERM_CONN`: The remote user terminated connection. Meaning the connection was explicitly disconnected e.g. HCI Disconnect commands was issued on remote part.
`BT_HCI_ERR_TERM_DUE_TO_MIC_FAIL`: The message integrity check failed meaning the encrypted PDU was corrupted.

There was a discussion on Jira: https://bluetooth.atlassian.net/jira/software/c/projects/ES/issues/ES-22692. I've added a comment there as I see no solution provided. There are 3 links linked but not accessible for me.

[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c:111]: I've had offline discussion with @sjanc, and we agreed that the ASE transition to QoS shall be for each CIS disconnection not preceded by ASE Release operation. I'll update this PR, so that the related PTS tests will pass

File: tests/bluetooth/audio/mocks/src/iso.c [hunk 2]
```diff
@@ -42,11 +42,7 @@ int bt_iso_server_unregister(struct bt_iso_server *server)
 
 int bt_iso_chan_disconnect(struct bt_iso_chan *chan)
 {
-	chan->state = BT_ISO_STATE_DISCONNECTED;
-	chan->ops->disconnected(chan, 0x13);
-	free(chan->iso);
-
-	return 0;
+	return mock_bt_iso_disconnected(chan, 0x13);
```
Comment (fredrikdanebjer) [inline tests/bluetooth/audio/mocks/src/iso.c:45]: Why are you using magic number here?

File: tests/bluetooth/audio/ascs/src/main.c [hunk 3]
```diff
@@ -475,3 +475,100 @@ ZTEST_F(ascs_test_suite, test_recv_in_enabling_state)
...
+	} else {
+		ase = fixture->ase_src.attr;
+		ase_id = fixture->ase_src.id;
+	}
+	zexpect_not_null(ase);
+	zexpect_true(ase_id != 0x00);
+
+	bt_bap_unicast_server_register_cb(&mock_bap_unicast_server_cb);
+
+	test_preamble_state_streaming(conn, ase_id, stream, &chan,
```
Comment (fredrikdanebjer) [inline tests/bluetooth/audio/ascs/src/main.c:499]: Should we also cover Source going from Streaming->Disabling->QoS here?

File: subsys/bluetooth/audio/ascs.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/ascs.c]: `BT_HCI_ERR_CONN_FAIL_TO_ESTAB` may be a special case though, as that will come if the CIS was not established in the first place, but it will still trigger a call to disconnected. If we are in the enabling state and we get `BT_HCI_ERR_CONN_FAIL_TO_ESTAB`, we should, IMO, stay in the enabling state.

[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c]: As per spec, the autonomous transition shall happen if ASE is in Disabling or Streaming state only

File: tests/bluetooth/audio/ascs/src/main.c (hunk 3 above)
Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c:499]: It's exactly what's covered in this particular scenario, however I've added second version of this test case Enabling->Disabling

File: subsys/bluetooth/audio/ascs.c [hunk 4]
```diff
@@ -897,18 +905,18 @@ static void ascs_ep_iso_disconnected(struct bt_bap_ep *ep, uint8_t reason)
...
 		 */
-		if (ep->status.state == BT_BAP_EP_STATE_STREAMING ||
-		    ep->status.state == BT_BAP_EP_STATE_ENABLING) {
-			if (ep->dir == BT_AUDIO_DIR_SOURCE) {
-				ascs_ep_set_state(ep, BT_BAP_EP_STATE_DISABLING);
-			} else {
-				ascs_ep_set_state(ep, BT_BAP_EP_STATE_QOS_CONFIGURED);
-			}
+		if (ep->dir == BT_AUDIO_DIR_SOURCE) {
+			ase->link_loss = true;
```
Comment (Thalley) [inline subsys/bluetooth/audio/ascs.c:913]: Suggest to move this assignment next to `ep->reason = reason;`. While it is currently only used during the BT_BAP_EP_STATE_DISABLING handler, it may be useful to access at some later point, where it may as well reflect the actual value, which is also that `ase->ep->reason` is only valid if `ase->link_loss`. 

Alternatively, if you don't agree with the above, then the field in the struct should have a comment stating that this field is only valid for source ASEs in the streaming state.

**Review by Thalley — CHANGES_REQUESTED** [review]
The changes seem sound, but I do have one small request

File: subsys/bluetooth/audio/ascs.c (hunk 4 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/ascs.c]: I was thinking about it for a while, and I think `link_loss` can only be set when ASE is in `BT_BAP_EP_STATE_STREAMING` or `BT_BAP_EP_STATE_DISABLING`.
Consider the because if set in e.g `BT_BAP_EP_STATE_ENABLING` and the server initiates Disable, we cannot go to QoS autonomously then. I updated the `test_cis_link_loss_in_enabling_state` test case to cover this scenario

**Review by Thalley — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
